builder/chroots.go

```diff
@@ -0,0 +1,637 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return cfg, nil
+}
+
+func (b *Builder) buildBundleChroots(set *bundleSet) error {
+	var err error
+	stateDir := b.Statedir
+	chrootDir := filepath.Join(stateDir, "image")
+
+	// TODO: Remove remaining references to outputDir. Let "build update" take care of
+	// bootstraping or cleaning up.
+	outputDir := filepath.Join(stateDir, "www")
+
+	if stateDir == "" || stateDir == "/" {
+		return errors.Errorf("invalid state dir: %s", stateDir)
+	}
+
+	if _, ok := set.Bundles["os-core"]; !ok {
+		return fmt.Errorf("os-core bundle not found")
+	}
+
+	// Bootstrap the directories.
+	err = os.MkdirAll(filepath.Join(chrootDir, "0"), 0755)
+	if err != nil {
+		return err
+	}
+	err = os.MkdirAll(filepath.Join(outputDir, "0"), 0755)
+	if err != nil {
+		return err
+	}
+	// TODO: Move this outside this function. chroot building doesn't use LAST_VER.
+	lastVerPath := filepath.Join(chrootDir, "LAST_VER")
+	if _, err = os.Stat(lastVerPath); os.IsNotExist(err) {
+		err = ioutil.WriteFile(lastVerPath, []byte("0\n"), 0644)
+		if err != nil {
+			return err
+		}
+	}
+
+	// TODO: To not touch config code that is in flux at the moment, reparsing it here to grab
+	// information that previously Mixer didn't care about. Move that to the configuration part
+	// of Mixer.
+	cfg, err := readBuildChrootsConfig(b.Buildconf)
+	if err != nil {
+		return err
+	}
+
+	if _, ok := set.Bundles[cfg.UpdateBundle]; !ok {
+		return fmt.Errorf("couldn't find bundle %q specified in configuration as the update bundle", cfg.UpdateBundle)
+	}
+
+	// Write INI files. These are used to communicate to the next step of mixing (build update).
+	var serverINI bytes.Buffer
+	fmt.Fprintf(&serverINI, `[Server]
+emptydir=%s/empty
+imagebase=%s/image/
+outputdir=%s/www/
+`, stateDir, stateDir, stateDir)
+	if cfg.HasServerSection {
+		fmt.Fprintf(&serverINI, `
+[Debuginfo]
+banned=%s
+lib=%s
+src=%s
+`, cfg.DebugInfoBanned, cfg.DebugInfoLib, cfg.DebugInfoSrc)
+	}
+	err = ioutil.WriteFile(filepath.Join(stateDir, "server.ini"), serverINI.Bytes(), 0644)
+	if err != nil {
+		return err
+	}
+	// TODO: If we are using INI files that are case insensitive, we need to be more restrictive
+	// in bundleset to check for that. See also readGroupsINI in swupd package.
+	var groupsINI bytes.Buffer
+	for _, bundle := range set.Bundles {
+		fmt.Fprintf(&groupsINI, "[%s]\ngroup=%s\n\n", bundle.Name, bundle.Name)
+	}
+	err = ioutil.WriteFile(filepath.Join(stateDir, "groups.ini"), groupsINI.Bytes(), 0644)
+	if err != nil {
+		return err
+	}
+
+	// Mixer is used to create both Clear Linux or a mix of it.
+	var version string
+	if b.Mixver != "" {
+		fmt.Printf("Creating chroots for version %s based on Clear Linux %s\n", b.Mixver, b.Clearver)
+		version = b.Mixver
+	} else {
+		fmt.Printf("Creating chroots for version %s\n", b.Clearver)
+		version = b.Clearver
+	}
+	// TODO: This validation should happen when reading the configuration.
+	if version == "" {
```

Comment:
Could add this in the else case above and remove from here.

Reply:
Done.